### ISSUE TITLE
Rook operator fails managing the cephclusters.ceph.rook.io CR on deletion

### ISSUE THREAD

**Issue by jparrill** [rook/rook#2944]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
The resources from cephclusters.ceph.rook.io CR stays after CR deletion 

**Expected behavior:**
Resources deletion when you delete the cephclusters.ceph.rook.io CR which include pods from this new cluster.

**How to reproduce it (minimal and precise):**
Deploy Ceph cluster following the openshift.md and the delete the CRD once all pods are on READY state.

**Environment**:
* OS (e.g. from /etc/os-release): CentOS Linux 7 (Core)
* Kernel (e.g. `uname -a`): 3.10.0-957.10.1.el7.x86_64
* Cloud provider or hardware configuration: Libvirt
* Rook version (use `rook version` inside of a Rook Pod): Master branch
* Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"4", Minor:"0+", GitVersion:"v4.0.22", GitCommit:"219bbe2f0c", GitTreeState:"", BuildDate:"2019-03-10T22:23:11Z", GoVersion:"", Compiler:"", Platform:""}
Server Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.4+d216153", GitCommit:"d216153", GitTreeState:"clean", BuildDate:"2019-04-07T09:36:47Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
```
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): OpenShift
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/Rook/master/toolbox.html)): `HEALTH_WARN Reduced data availability: 100 pgs inactive; clock skew detected on mon.b, mon.a`

* Reproducing the status:

```
NAME        DATADIRHOSTPATH   MONCOUNT   AGE   STATE
rook-ceph   /var/lib/rook     3          12m   Created
```

```
[jparrill@sjr3 ceph]$ oc delete cephclusters.ceph.rook.io rook-ceph
cephcluster.ceph.rook.io "rook-ceph" deleted
```

After delete the CRD the resources stays there:

```
[jparrill@sjr3 dev-scripts]$ oc get pods
NAME                                   READY   STATUS             RESTARTS   AGE
rook-ceph-agent-dgp75                  1/1     Running            0          50m
rook-ceph-agent-kspf7                  1/1     Running            0          50m
rook-ceph-agent-qrshb                  1/1     Running            0          50m
rook-ceph-mgr-a-7b8cd8c88b-v5g9b       1/1     Running            0          39m
rook-ceph-mon-a-cfc6799fd-dzn49        1/1     Running            0          41m
rook-ceph-mon-b-65df486949-klq6g       1/1     Running            0          40m
rook-ceph-mon-c-6c8597d468-x2bdj       1/1     Running            0          40m
rook-ceph-operator-654b5c69b5-g5zgv    1/1     Running            0          50m
rook-ceph-osd-0-54c9d6f7cf-vl5gq       0/1     CrashLoopBackOff   4          12m
rook-ceph-osd-1-9d6b75d9f-mz8hc        0/1     CrashLoopBackOff   4          11m
rook-ceph-osd-2-6d94489d88-d4w7b       0/1     CrashLoopBackOff   4          12m
rook-ceph-osd-prepare-master-0-wdrtd   0/2     Completed          0          12m
rook-ceph-osd-prepare-master-1-ksl48   0/2     Completed          0          12m
rook-ceph-osd-prepare-master-2-c88mv   0/2     Completed          0          12m
rook-discover-99jr6                    1/1     Running            0          50m
rook-discover-b28cv                    1/1     Running            0          50m
rook-discover-j6752                    1/1     Running            0          50m
```

**Comment by jparrill** [comment]
More details about this bug:

The deployment of operator and cluster was fine, but when rook-ceph-osd-prepare-master pods appear does not detect right (or this is what I see) the free disk. This is the log of the osd-prepare pod:

```
2019-04-08 15:13:20.283176 I | cephosd: skipping device vda that is in use (not by rook). fs: , ownPartitions: false
2019-04-08 15:13:20.292280 I | cephosd: configuring osd devices: {"Entries":{}}
```

If you go to the master node you could see this:
```
NAME                                                                                                 MAJ:MIN RM SIZE RO TYPE MOUNTPOINT
vda                                                                                                  252:0    0  50G  0 disk 
├─vda1                                                                                               252:1    0   1M  0 part 
├─vda2                                                                                               252:2    0   1G  0 part /boot
├─vda3                                                                                               252:3    0  49G  0 part /sysroot
└─vda4                                                                                               252:4    0  65M  0 part 
vdb                                                                                                  252:16   0  70G  0 disk 
└─ceph--98e27435--4435--4833--b633--fcd8db81aefc-osd--data--39614815--114a--4f34--9451--3637441a0e81 253:0    0  70G  0 lvm  
```

This does not make sense for me, if you cannot detect a disk why did you prepare the lvm over it?.... well the oher elements goes up....mon, mgr, osds....but yeah osds are going up and down every 2 minutes and the cluster cannot be provisioned:

```
rook-ceph-osd-0-5d8fb5f647-r2ldt       0/1     CrashLoopBackOff   5          14m
rook-ceph-osd-1-74f99ff999-lthbp       0/1     CrashLoopBackOff   5          14m
rook-ceph-osd-2-54476867d4-hs8fl       0/1     Error              5          14m
```

With this error:

```
2019-04-08 15:26:05.082538 I | exec: Running command: ceph-volume lvm activate --no-systemd --bluestore 0 af80324a-a71c-4f35-9ea2-cc215ec54edb
2019-04-08 15:27:59.777923 I | Running command: /bin/mount -t tmpfs tmpfs /var/lib/ceph/osd/ceph-0
2019-04-08 15:27:59.778547 I | Running command: /usr/sbin/restorecon /var/lib/ceph/osd/ceph-0
2019-04-08 15:27:59.778620 I | Running command: /bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-0
2019-04-08 15:27:59.778694 I | Running command: /bin/ceph-bluestore-tool --cluster=ceph prime-osd-dir --dev /dev/ceph-459b5db7-16d8-4ab2-8217-021f69d7a64d/osd-data-c637f30f-604e-4993-a530-56872ebb757f --path /var/lib/ceph/osd/ceph-0 --no-mon-config
2019-04-08 15:27:59.778834 I | Running command: /bin/ln -snf /dev/ceph-459b5db7-16d8-4ab2-8217-021f69d7a64d/osd-data-c637f30f-604e-4993-a530-56872ebb757f /var/lib/ceph/osd/ceph-0/block
2019-04-08 15:27:59.778912 I | Running command: /bin/chown -h ceph:ceph /var/lib/ceph/osd/ceph-0/block
2019-04-08 15:27:59.778976 I | Running command: /bin/chown -R ceph:ceph /dev/dm-0
2019-04-08 15:27:59.779039 I | Running command: /bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-0
2019-04-08 15:27:59.779106 I | --> ceph-volume lvm activate successful for osd ID: 0
2019-04-08 15:27:59.790169 I | exec: Running command: ceph-osd --foreground --id 0 --osd-uuid af80324a-a71c-4f35-9ea2-cc215ec54edb --conf /var/lib/rook/osd0/rook-ceph.config --cluster ceph
```
Ideas?

**Comment by travisn** [comment]
@jparrill In the Rook operator log, can you search to see if the following message is present?
```
OwnerReferences will not be set on resources created by rook. failed to test that it can be set.
```

If so, there is still an issue with setting the OwnerReferences in OpenShift when the owner is a custom resource. Rook intentionally skips setting the owner refs when they fail [here](https://github.com/rook/rook/blob/master/pkg/operator/k8sutil/resources.go#L90-L110). If that is the case, [this k8s PR](https://github.com/kubernetes/kubernetes/pull/62810) must not have fixed the issue as expected.

Without the capability of the OwnerReferences, cleanup of the rook resources would need to be done by deleting the namespace, or otherwise deleting all the resources in the rook namespace.

**Comment by jparrill** [comment]
Ahh ok, there it is:

```
2019-04-08 11:31:44.687572 W | op-k8sutil: OwnerReferences will not be set on resources created by rook. failed to test that it can be set. configmaps "rook-test-ownerref" is forbidden: cannot set blockOwnerDeletion in this case because cannot find RESTMapping for APIVersion v1 Kind CephCluster: no matches for kind "CephCluster" in version "v1"
2019-04-08 11:31:44.713025 I | op-k8sutil: waiting for job rook-ceph-detect-version to complete...
2019-04-08 11:31:59.744745 E | op-cluster: unknown ceph major version. failed to get version job log to detect version. failed to read from stream. Get https://master-0:10250/containerLogs/rook-ceph/rook-ceph-detect-version-vpzgd/version: remote error: tls: internal error
```

But I am using openshift 4 master (after the 1.13 k8s rebase) branch over libvirt, I think that the code must be there, isn't it?

**Comment by travisn** [comment]
@jparrill Correct, we certainly must have the latest merges that I had referenced where this was supposedly fixed and my only explanation is that it isn't fixed and we haven't followed up on that yet. This has always been working in Kubernetes, but not OpenShift. It would be very helpful if you could follow up with OpenShift on this. For the repro, it should be straight forward to see that any resource (for example ConfigMap) that is created will fail if it has an OwnerReference that is a custom resource (for example CephCluster).

**Comment by travisn** [comment]
Following up on comments in #1533...

**Comment by travisn** [comment]
Reopening to consider for backport to 1.0.

**Comment by travisn** [comment]
Will be included in v1.0.3
